### ISSUE TITLE
RelatedObjectDoesNotExist exception on row_result.object_repr attribute initialization

### ISSUE THREAD

**Issue by kygoh** [django-import-export/django-import-export#1556]
Extending the example from #439, I have the following 3 models:

```py
class Candidate(models.Model):
    first_name = models.CharField("First name", max_length=255)

class Test(models.Model):
    name = models.CharField("Test name", max_length=255)

    def __str__(self):
        return self.name

class CandidateTest(models.Model):
    candidate = models.ForeignKey(Candidate, on_delete=models.CASCADE)
    test = models.ForeignKey(Test, on_delete=models.CASCADE)

    def __str__(self):
        return '{} taking {}'.format(self.candidate.first_name, self.test.name)
```

and the following import resource class:

```py
class CandidateTestResource(ModelResource):
    candidate = fields.Field(attribute='candidate', column_name='candidate', widget=CharWidget())
    test = fields.Field(attribute='test', column_name='test', widget=CharWidget())

    class Meta:
        model = CandidateTest

    def import_field(self, field, instance, row, is_m2m=False, **kwargs):
        field_name = self.get_field_name(field)
        method = getattr(self, 'clean_%s' % field_name, None)
        if method is not None:
            instance = method(field, instance, row, **kwargs)
        else:
            super().import_field(field, instance, row, is_m2m, **kwargs)

    def clean_candidate(self, field, instance, row, **kwargs):
        if row['candidate']:
            try:
                candidate, _ = Candidate.objects.get_or_create(
                    first_name=row['candidate']
                )
                instance.candidate = candidate
            except:
                pass

        return instance

    def clean_test(self, field, instance, row, **kwargs):
        if row['test']:
            try:
                test, _ = Test.objects.get_or_create(
                    name=row['test']
                )
                instance.test = test
            except:
                pass

        return instance

    def skip_row(self, instance, original, row, import_validation_errors=None):
        if not (row['candidate'] and row['test']):
            return True

        return super().skip_row(instance, original, row, import_validation_errors=import_validation_errors)
```

The following csv is used to import:

```csv
id,candidate,test
1,Candidate 1,Test A
2,Candidate 2,
3,,Test B
```

The above code works as expected for Django Import Export 3.0.1:

<img width="316" alt="Screenshot 2023-03-02 at 12 50 35 AM" src="https://user-images.githubusercontent.com/5347201/222207301-7048c36c-8a92-4754-93e1-eadf59ba7961.png">

but throws the following error from version 3.0.2 onwards:

```
Line number: 2 - CandidateTest has no test.
2, Candidate 2,
Traceback (most recent call last):
File "../import_export/resources.py", line 750, in import_row
row_result.add_instance_info(instance)
File "../import_export/results.py", line 43, in add_instance_info
self.object_repr = force_str(instance)
File "/home/ubuntu/.local/lib/python3.6/site-packages/django/utils/encoding.py", line 64, in force_str
s = str(s)
File "/home/ubuntu/django-import-export/tests/core/models.py", line 197, in __str__
return '{} taking {}'.format(self.candidate.first_name, self.test.name)
File "/home/ubuntu/.local/lib/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 198, in __get__
"%s has no %s." % (self.field.model.__name__, self.field.name)
core.models.CandidateTest.test.RelatedObjectDoesNotExist: CandidateTest has no test.

Line number: 3 - CandidateTest has no candidate.
3, , Test B
Traceback (most recent call last):
File "../import_export/resources.py", line 750, in import_row
row_result.add_instance_info(instance)
File "../import_export/results.py", line 43, in add_instance_info
self.object_repr = force_str(instance)
File "/home/ubuntu/.local/lib/python3.6/site-packages/django/utils/encoding.py", line 64, in force_str
s = str(s)
File "/home/ubuntu/django-import-export/tests/core/models.py", line 197, in __str__
return '{} taking {}'.format(self.candidate.first_name, self.test.name)
File "/home/ubuntu/.local/lib/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 198, in __get__
"%s has no %s." % (self.field.model.__name__, self.field.name)
core.models.CandidateTest.candidate.RelatedObjectDoesNotExist: CandidateTest has no candidate.
```
The suggested modification from https://github.com/django-import-export/django-import-export/issues/439#issuecomment-215691637 is required again to resolve the error in 3.0.2 onwards:
```py
class CandidateTest(models.Model):
    candidate = models.ForeignKey(Candidate, on_delete=models.CASCADE)
    test = models.ForeignKey(Test, on_delete=models.CASCADE)

    def __str__(self):
        try:
            return '{} taking {}'.format(self.candidate.first_name, self.test.name)
        except ObjectDoesNotExist as err: 
            return '[Invalid instance - %s]' % err
```

**Comment by matthewhegarty** [comment]
For reference, this new behaviour was introduced in #1526 (where an instance is added to a Result), which then has a call to `str(instance)`.

**Comment by matthewhegarty** [comment]
If it is possible that `CandidateTest.test` can be null, surely this should be handled in the `__str__()` method (as per this [comment](https://github.com/django-import-export/django-import-export/issues/439#issuecomment-242353786)), and also in your implementation of `__str__()`.

Surely this is [best practice](https://docs.python.org/3/glossary.html#term-eafp) and not a bug in import-export.  Am I missing something?

**Comment by kygoh** [comment]
> If it is possible that `CandidateTest.test` can be null,

Technically `CandidateTest.test` cannot be null as `null=True` is not set:
```
test = models.ForeignKey(Test, on_delete=models.CASCADE)
```

> Am I missing something?

`CandidateTest` is a junction table for many to many relationship between `Candidate` and `Test`.

It is actually an attempt to consolidate 3 different import steps into one i.e.:

1. Import `Candidate`
2. Import `Test`
3. Import `CandidateTest`

Thus, the use case for this import is to:

1. Create (if not exists) `Candidate` who has not registered for any `Test`, if only `candidate` column has value
2. Create (if not exists) `Test` that was newly introduced, if only `test` column has value
3. Create (if not exists) `Candidate`, `Test` then relate the `Candidate` taking the `Test` in `CandidateTest`, if both columns have values

**Comment by kygoh** [comment]
> For reference, this new behaviour was introduced in #1526 (where an instance is added to a Result)

I am trying to figure out #1526:

> The RowResults in the case of skipped rows during import do not include the skipped instance's information.
> There are several use-cases when the primary keys of the imported rows are needed. 

I assume skipped rows will not be imported thus will not have primary keys and not logged: :
https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/admin.py#L207-L215

So this test makes little sense:
https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/tests/core/tests/test_resources.py#L1144-L1146

**Comment by kygoh** [comment]
> surely this should be handled in the `__str__()` method (as per this [comment](https://github.com/django-import-export/django-import-export/issues/439#issuecomment-242353786))

https://github.com/django-import-export/django-import-export/issues/459#issuecomment-219687467:
> As it seems, the models __str__ or __unicode__ method is used as an attribute of the row_result object, but I think this only internal and never presented to the user, so maybe a pull request would be good idea. Instead of using try ... except in the models method, it could be used directly in the script to use a placeholder if row_result.object_repr = force_text(instance) throws an error. But maybe row_result.object_repr has another important function that I'm missing here.

https://github.com/django-import-export/django-import-export/issues/493#issuecomment-234025420

> You can except the error in your model's  `__unicode__` method as in #439. However it's probably better to only except the specific error instead of using the broad except with
>
> ``` python
> from django.core.exceptions import DoesNotExist
> ```
>
> Yet I think this is an issue with `django-import-export` and should be fixed at some point.

I think it was deemed an issue that required a fixed in #505:

> I only write the properties when the row has been fully materialized, so the FKs are already there after saving.
>
> Also, if we skip we don't write the LogEntry, so no need to write the properties.
>
> Closes https://github.com/django-import-export/django-import-export/issues/493, https://github.com/django-import-export/django-import-export/issues/439, https://github.com/django-import-export/django-import-export/issues/459

**Comment by kygoh** [comment]
The `instance` in:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L750

originates from either: (1) an already existing instance:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L375

or (2) a newly initialized but _empty_ instance:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L378

For (2), the _empty_ instance will be populated with data from `row`:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L737

then validated and saved:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L747-L749

However, if the row is skipped:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L744

there is a possibility the `instance` is not (or partially) populated and still invalid.

Since `instance` is still invalid, calling

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/resources.py#L750

will raise the reported `RelatedObjectDoesNotExist`.

I think that exception should be caught:

https://github.com/django-import-export/django-import-export/blob/4ac05ae6028555f1674df32415b3d4e2afbf8720/import_export/results.py#L39-L43

**Comment by matthewhegarty** [comment]
Thanks for raising this and for doing the additional research into the problem.  I agree that the change in #1526 could cause regression issues for implementations where str() is overridden to and references can be null.  However I do think this is a client issue rather than a library issue.

I don't think it's good practice to catch and ignore these exceptions as in #1559.  I can't get past the idea that the exception is coming from external user-defined code which we have no control over.  I think for client defined `__str__()` methods it would be best practice to handle any potential null references.

If any exceptions are raised from `add_instance_info()` then these are already [caught and handled](https://github.com/django-import-export/django-import-export/blob/54bc71662050a2a57da8fd6e31c8afd87f4dd48f/import_export/resources.py#L762). 

My preference is to make no additional changes to the library, and therefore close the issue and PR.  However, this issue will serve as a reference should anyone have similar issues in future.

Thanks again for your contribution and for looking into this.

**Comment by kygoh** [comment]
I truly appreciate your acknowledgement and responsiveness, also not forgetting your effort in maintaining this helpful library.

I intuitively assumed the library will just ignore a row because I wanted to skip it but find it confusing to suddenly get a "RelatedObjectDoesNotExist" exception. Having gone through issues and PR raised previously (#439, #459, #505, #508), I think they had used similar assumption and were similarly confounded.

Originally `add_instance_info` is called under 2 conditions:
1. row is imported
2. row is deleted

Under both conditions, `instance` is a model object already stored in the database. Thus, calling `add_instance_info` makes sense as it conforms to the context stated in [Django's documentation](https://docs.djangoproject.com/en/4.1/ref/models/instances/#str):

> display an object in the Django admin site and as the value inserted into a template when it displays an object.

However, #1526 calls `add_instance_info` even when `instance` is not a model object stored in the database. I think this is the wrong context to call `str`, thus a library issue. Even `add_instance_info` has to handle cases when `instance` is not a model object stored in the database:

https://github.com/django-import-export/django-import-export/blob/54bc71662050a2a57da8fd6e31c8afd87f4dd48f/import_export/results.py#L42

Initially, I wanted to change `add_instance_info` to:

```
        if instance is not None and instance.pk is not None:
            # Add object info to RowResult (e.g. for LogEntry)
            self.object_id = instance.pk
            self.object_repr = force_str(instance)
```

but since you mentioned [EAFP](https://docs.python.org/3/glossary.html#term-EAFP) as best practice I submitted #1559 instead.

As a recent user of the Django framework, I think the reason Django and many of its related libraries live up to "the web framework for perfectionists with deadlines" tagline is because most of the implementation details are shielded from the developer. I feel that it involves more configuration rather than coding.

Getting a "RelatedObjectDoesNotExist" exception and having to go back to my models `__str__` code to handle it (which I think may be numerous for some) may be exposing too much library implementation details.

**Comment by matthewhegarty** [comment]
OK - thanks for the comments - I take your points.

To summarise the issue:

1.  If you have overridden `__str__()` to reference a persisted instance then this can cause a crash when `add_instance_info()` is called.  This is because `force_str()` is called on the instance.  It happens because the reference may not exist at the time of import, either in the case of 'new' instances, or FK relations which don't exist yet.  

2.  The original code did not store call `force_str()` for skipped rows ([505](https://github.com/django-import-export/django-import-export/pull/505/files#diff-c2f872217bf726f1498a762c2e74c1daa2c19fe17a747d341c5d679d26ffbb0eR461)) but this was changed in #1526.  It is the case that `skip_row()` can be overridden and can therefore operate on both existing or new instances.

3.  We have had similar issues raised before and fixes have been supplied (as referenced earlier in thread).  It's likely that this will be raised as an issue again.

I'm still of the opinion that `__str__()` implementations should be handling null references, but I guess there are cases where this cannot easily be done (maybe legacy code in a Django app).  Users might not be happy with having to change existing code to get it to work with this library.

Also, I don't think that catching and ignoring exceptions is a good idea.  If we did this, then we would likely get complaints from people who spend hours debugging only to find that we are swallowing exceptions.  An exception in `__str__()` could be raised for any reason which is completely outside of our control.

I suggest the following as a workaround.  The library supports writing a custom `RowResult` which would handle this issue.  For example:

```python
class MyRowResult(RowResult):
    def add_instance_info(self, instance):
        try:
            super().add_instance_info(instance)
        except:
            pass
```

Then reference this in your Resource:

```python
class BookResource(ModelResource):

    class Meta:
        model = Book

    @classmethod
    def get_row_result_class(self):
        return MyRowResult
```

I think this is a pragmatic solution for what is mostly an edge case.  I will note this in the FAQ with a link to this thread.  If we get other cases of this occurring then perhaps we can consider a change to the library to accommodate.
